xmr-btc/tests/e2e.rs

```diff
@@ -0,0 +1,386 @@
+        alice_xmr: alice_initial_xmr_balance,
+        alice_btc: alice_initial_btc_balance,
+        bob_xmr: bob_initial_xmr_balance,
+        bob_btc: bob_initial_btc_balance,
+    };
+    (
+        alice_state0,
+        bob_state0,
+        alice,
+        bob,
+        initial_balances,
+        swap_amounts,
+    )
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::{
+        harness::node::{run_alice_until, run_bob_until},
+        init_bitcoind, init_test,
+    };
+
+    use monero_harness::Monero;
+    use rand::rngs::OsRng;
+    use testcontainers::clients::Cli;
+
+    use futures::future;
+    use std::convert::TryInto;
+    use tracing_subscriber::util::SubscriberInitExt;
+    use xmr_btc::{
+        alice, bitcoin,
+        bitcoin::{Amount, TX_FEE},
+        bob,
+    };
+
+    #[tokio::test]
+    async fn happy_path() {
+        let _guard = tracing_subscriber::fmt()
+            .with_env_filter("info")
+            .set_default();
+
+        let cli = Cli::default();
+        let monero = Monero::new(&cli);
+        let bitcoind = init_bitcoind(&cli).await;
+
+        let (
+            alice_state0,
+            bob_state0,
+            mut alice_node,
+            mut bob_node,
+            initial_balances,
+            swap_amounts,
+        ) = init_test(&monero, &bitcoind).await;
+
+        let (alice_state, bob_state) = future::try_join(
+            run_alice_until(
+                &mut alice_node,
+                alice_state0.into(),
+                alice::is_state6,
+                &mut OsRng,
+            ),
+            run_bob_until(&mut bob_node, bob_state0.into(), bob::is_state5, &mut OsRng),
+        )
+        .await
+        .unwrap();
+
+        let alice_state6: alice::State6 = alice_state.try_into().unwrap();
+        let bob_state5: bob::State5 = bob_state.try_into().unwrap();
+
+        let alice_final_btc_balance = alice_node.bitcoin_wallet.balance().await.unwrap();
+        let bob_final_btc_balance = bob_node.bitcoin_wallet.balance().await.unwrap();
+
+        let lock_tx_bitcoin_fee = bob_node
+            .bitcoin_wallet
+            .transaction_fee(bob_state5.tx_lock_id())
+            .await
+            .unwrap();
+
+        assert_eq!(
+            alice_final_btc_balance,
+            initial_balances.alice_btc + swap_amounts.btc
+                - bitcoin::Amount::from_sat(bitcoin::TX_FEE)
+        );
+        assert_eq!(
+            bob_final_btc_balance,
+            initial_balances.bob_btc - swap_amounts.btc - lock_tx_bitcoin_fee
+        );
+
+        let alice_final_xmr_balance = alice_node
+            .monero_wallet
+            .0
+            .get_balance_alice()
+            .await
+            .unwrap();
+
+        bob_node
+            .monero_wallet
+            .0
+            .wait_for_bob_wallet_block_height()
+            .await
+            .unwrap();
+
+        let bob_final_xmr_balance = bob_node.monero_wallet.0.get_balance_bob().await.unwrap();
+
```

Comment:
If possible, I would always group the asserts together in the end. That makes it easier to follow that the test is now over once you reach the assertions.

```suggestion
        let alice_final_xmr_balance = alice_node
            .monero_wallet
            .0
            .get_balance_alice()
            .await
            .unwrap();

        bob_node
            .monero_wallet
            .0
            .wait_for_bob_wallet_block_height()
            .await
            .unwrap();

        let bob_final_xmr_balance = bob_node.monero_wallet.0.get_balance_bob().await.unwrap();

        assert_eq!(
            alice_final_btc_balance,
            initial_balances.alice_btc + swap_amounts.btc
                - bitcoin::Amount::from_sat(bitcoin::TX_FEE)
        );
        assert_eq!(
            bob_final_btc_balance,
            initial_balances.bob_btc - swap_amounts.btc - lock_tx_bitcoin_fee
        );
```